datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -743,9 +714,6 @@ fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataTyp
     use arrow::datatypes::IntervalUnit::*;
     use arrow::datatypes::TimeUnit::*;
 
-    if lhs_type == rhs_type {
```

Comment:
This was incorrect if the types are not temporal, rather than adding a check, I just removed it as we explicitly test the types below